File: src/main/scala/committee/nova/skillful/player/capabilities/impl/Skills.scala
package committee.nova.skillful.player.capabilities.impl

import committee.nova.skillful.api.skill.ISkill
import committee.nova.skillful.event.handler.CapabilityHandler.skillfulCap
import committee.nova.skillful.impl.skill.instance.SkillInstance
import committee.nova.skillful.implicits.Implicits.PlayerEntityImplicit
import committee.nova.skillful.player.capabilities.api.ISkills
import committee.nova.skillful.player.capabilities.info.SkillInfo
import committee.nova.skillful.storage.SkillfulStorage
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.{CompoundNBT, INBT, ListNBT}
import net.minecraft.util.{Direction, ResourceLocation}
import net.minecraftforge.common.capabilities.Capability.IStorage
import net.minecraftforge.common.capabilities.{Capability, ICapabilityProvider, ICapabilitySerializable}
import net.minecraftforge.common.util.{LazyOptional, NonNullSupplier}

import scala.collection.mutable
import scala.jdk.CollectionConverters._

object Skills {
  class Provider extends ICapabilityProvider with ICapabilitySerializable[CompoundNBT] {
    private val instance = new Impl

    private val storage = new Storage

    override def serializeNBT(): CompoundNBT = {
      val tag = new CompoundNBT
      tag.put("skillful", storage.writeNBT(skillfulCap, instance, null))
      tag
    }

    override def deserializeNBT(nbt: CompoundNBT): Unit = storage.readNBT(skillfulCap, instance, null, nbt.getList("skillful", 10))

    override def getCapability[T](cap: Capability[T], side: Direction): LazyOptional[T] = if (cap == skillfulCap) LazyOptional.of(new NonNullSupplier[ISkills] {
      override def get(): ISkills = instance
    }.asInstanceOf[NonNullSupplier[T]]) else null

  }

  class Storage extends IStorage[ISkills] {
    override def writeNBT(capability: Capability[ISkills], instance: ISkills, side: Direction): INBT = {
      val tag = new ListNBT()
      instance.getSkills.foreach(s => tag.add(s.serializeNBT()))
      tag
    }

    override def readNBT(capability: Capability[ISkills], instance: ISkills, side: Direction, nbt: INBT): Unit = {
      nbt match {
        case tag: ListNBT =>
          val skills = instance.getSkills
          skills.clear()
          tag.asScala.foreach {
            case t: CompoundNBT =>
              val skill = SkillfulStorage.getSkill(new ResourceLocation(t.getString("skill")))
              val i = new SkillInstance(skill)
              i.deserializeNBT(t)
              skills.add(i)
            case _ =>
          }
        case _ =>
      }
    }
  }

  class Impl extends ISkills {
    private val skills: mutable.HashSet[SkillInstance] = new mutable.HashSet[SkillInstance]()

    private val skillInfos: mutable.HashSet[SkillInfo] = new mutable.HashSet[SkillInfo]()

    override def getSkills: mutable.HashSet[SkillInstance] = skills

    override def getSkill(skill: ISkill): SkillInstance = {
      skills.foreach(s => if (s.getSkill.equals(skill)) return s)
      val instance = new SkillInstance(skill)
      skills.add(instance)
      instance
    }

    override def getSkillCleanly(skill: ISkill): Option[SkillInstance] = {
      skills.foreach(s => if (s.getSkill.equals(skill)) return Some(s))
      None
    }

    override def getSkillInfos: mutable.HashSet[SkillInfo] = skillInfos

    override def getSkillInfo(player: PlayerEntity, skill: ISkill): SkillInfo = {
      skillInfos.foreach(i => if (i.getSkillInstance.getSkill.equals(skill)) return i)
      val instance = new SkillInfo(player.getSkillStat(skill))
      skillInfos.add(instance)
      instance
    }

    override def removeSkill(id: ResourceLocation): Boolean = {
      skills.foreach(s => if (id.equals(s.getSkill.getId)) return skills.remove(s))
      false
    }
  }
}

File: src/main/java/committee/nova/skillful/common/cap/skill/Skills.java
package committee.nova.skillful.common.cap.skill;

import committee.nova.skillful.common.manager.SkillTypeManager;
import committee.nova.skillful.common.skill.DummySkillInstance;
import committee.nova.skillful.common.skill.ISkillType;
import committee.nova.skillful.common.skill.SkillInstance;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.minecraftforge.common.capabilities.CapabilityToken;
import net.minecraftforge.common.capabilities.ICapabilitySerializable;
import net.minecraftforge.common.util.LazyOptional;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Skills {
    public static final Capability<ISkills> SKILLS_CAPABILITY = CapabilityManager.get(new CapabilityToken<>() {
    });

    public static class Impl implements ISkills {
        private final Map<ISkillType, SkillInstance> skills = new HashMap<>();
        private final Map<ResourceLocation, DummySkillInstance> unregisteredSkills = new HashMap<>();

        @Override
        public Collection<SkillInstance> getSkills() {
            return skills.values();
        }

        @Override
        public boolean hasSkill(ISkillType id) {
            return skills.containsKey(id);
        }

        @Override
        public Optional<SkillInstance> getSkill(ISkillType skill) {
            return Optional.ofNullable(skills.get(skill));
        }

        @Override
        public SkillInstance getOrCreateSkill(ISkillType skill) {
            if (hasSkill(skill)) return skills.get(skill);
            final SkillInstance instance = SkillInstance.of(skill);
            skills.put(skill, instance);
            return instance;
        }

        @Override
        public Optional<DummySkillInstance> getUnregisteredSkill(ResourceLocation id) {
            return Optional.ofNullable(unregisteredSkills.get(id));
        }

        @Override
        public ListTag serializeNBT() {
            final ListTag list = new ListTag();
            skills.values().forEach(i -> list.add(i.serializeNBT()));
            unregisteredSkills.values().forEach(i -> list.add(i.serializeNBT()));
            return list;
        }

        @Override
        public void deserializeNBT(ListTag list) {
            skills.clear();
            unregisteredSkills.clear();
            list.forEach(i -> {
                if (!(i instanceof CompoundTag t)) return;
                final ResourceLocation id = ResourceLocation.tryParse(t.getString("skill"));
                if (id == null) return;
                if (SkillTypeManager.hasSkillType(id)) SkillTypeManager.getSkillType(id).ifPresent(s -> {
                    final SkillInstance instance = SkillInstance.of(s);
                    instance.deserializeNBT(t);
                    skills.put(s, instance);
                });
                else {
                    final DummySkillInstance dummy = DummySkillInstance.of(id);
                    dummy.deserializeNBT(t);
                    unregisteredSkills.put(id, dummy);
                }
            });
        }
    }

    public static class Provider implements ICapabilitySerializable<ListTag> {
        private ISkills skills;

        @Override
        public @NotNull <T> LazyOptional<T> getCapability(@NotNull Capability<T> cap, @Nullable Direction side) {
            return cap == SKILLS_CAPABILITY ? LazyOptional.of(this::getOrCreateCapability).cast() : LazyOptional.empty();
        }

        public ISkills getOrCreateCapability() {
            if (skills == null) {
                this.skills = new Impl();
            }
            return this.skills;
        }

        @Override
        public ListTag serializeNBT() {
            return getOrCreateCapability().serializeNBT();
        }

        @Override
        public void deserializeNBT(ListTag tag) {
            getOrCreateCapability().deserializeNBT(tag);
        }
    }
}
